File: MyDailyBeat/MyDailyBeat/EVCGroupSettingsViewController.m
//
//  EVCGroupSettingsViewController.m
//  MyDailyBeat
//
//  Created by Virinchi Balabhadrapatruni on 10/26/14.
//  Copyright (c) 2014 eVerveCorp. All rights reserved.
//

#import "EVCGroupSettingsViewController.h"

@interface EVCGroupSettingsViewController ()

@end

@implementation EVCGroupSettingsViewController

- (id) initWithGroup:(Group *) group andCompletionBlock:(EVCGroupSettingsViewControllerCompletionHandler) completion {
    self = [self initWithNibName:@"EVCGroupSettingsViewController_iPhone" bundle:nil];
    if (self) {
        _g = group;
        _handler = completion;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.formController = [[FXFormController alloc] init];
    self.formController.tableView = self.tableView;
    self.formController.delegate = self;
    GroupPrefs *prefs = [[GroupPrefs alloc] initWithServingURL:self.g.servingURL];
    prefs.hobbies = [[NSArray alloc] initWithArray:self.g.hobbies];
    self.formController.form = prefs;
    self.api = [RestAPI getInstance];
    
    UIBarButtonItem *cancelBarButton = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemCancel target:self action:@selector(cancel)];
    self.navigationItem.leftBarButtonItem = cancelBarButton;

    
    UIBarButtonItem *postBarButton = [[UIBarButtonItem alloc] initWithTitle:@"Done" style:UIBarButtonItemStyleDone target:self action:@selector(done)];
    self.navigationItem.rightBarButtonItem = postBarButton;


}

- (void) cancel {
    [self dismissViewControllerAnimated:YES completion:nil];
}



- (void) done {
    [self dismissViewControllerAnimated:YES completion:^{
        // save group preferences
        GroupPrefs *prefs = self.formController.form;
        
        if ([prefs.hobbies count] > 3) {
            [self.view makeToast:@"Cannot select more than 3 hobbies" duration:3.5 position:@"bottom"];
            return;
        }
        
        dispatch_queue_t queue = dispatch_queue_create("dispatch_queue_t_dialog", NULL);
        dispatch_async(queue, ^{
            dispatch_async(dispatch_get_main_queue(), ^{
                [self.view makeToastActivity];
            });
            self.g.hobbies = [[NSMutableArray alloc] initWithArray:prefs.hobbies];
            BOOL success = [self.api setHobbiesforGroup:self.g];
            dispatch_async(dispatch_get_main_queue(), ^{
                [self.view hideToastActivity];
                if (success) {
                    [self.view makeToast:@"Upload successful!" duration:3.5 position:@"bottom" image:[UIImage imageNamed:@"check.png"]];
                    _handler();
                } else {
                    [self.view makeToast:@"Upload failed!" duration:3.5 position:@"bottom" image:[UIImage imageNamed:@"error.png"]];
                    return;
                }
                
            });
            
        });
        
    }];
}

- (void) viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    [self loadGroupPicture];
    [self.tableView reloadData];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void) loadGroupPicture {
    dispatch_queue_t queue = dispatch_queue_create("dispatch_queue_t_dialog", NULL);
    dispatch_async(queue, ^{
        NSURL *imageURL = [[RestAPI getInstance] retrieveGroupPictureForGroup:self.g];
        if (imageURL == nil) return;
        NSData *imageData = [[RestAPI getInstance] fetchImageAtRemoteURL:imageURL];
        dispatch_async(dispatch_get_main_queue(), ^{
            // Update the UI
            GroupPrefs *prefs = self.formController.form;
             prefs.groupPicture = [UIImage imageWithData:imageData];
            
        });
    });
    
}

- (void)saveImage:(UITableViewCell<FXFormFieldCell> *)cell {
    GroupPrefs *prefs = self.formController.form;
    
    dispatch_queue_t queue = dispatch_queue_create("dispatch_queue_t_dialog", NULL);
    dispatch_async(queue, ^{
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.view makeToastActivity];
        });
        NSData *imgData = UIImagePNGRepresentation(prefs.groupPicture);
        
        NSString *fileName = ASSET_FILENAME;
        
        BOOL success = [[RestAPI getInstance] uploadGroupPicture:imgData withName:fileName toGroup:self.g];
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.view hideToastActivity];
            if (success)
                [self.view makeToast:@"Upload successful!" duration:3.5 position:@"bottom" image:[UIImage imageNamed:@"check.png"]];
            else {
                [self.view makeToast:@"Upload failed!" duration:3.5 position:@"bottom" image:[UIImage imageNamed:@"error.png"]];
                return;
            }
            
        });
        
    });
}

- (void)deleteGroup:(UITableViewCell<FXFormFieldCell> *)cell {
    dispatch_queue_t queue = dispatch_queue_create("dispatch_queue_t_dialog", NULL);
    dispatch_async(queue, ^{
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.view makeToastActivity];
        });
        
        BOOL success = [[RestAPI getInstance] deleteGroup:self.g];
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.view hideToastActivity];
            if (success)
                [self.delegate EVCGroupSettingsViewControllerDelegateDidDeleteGroup:self];
            else
                NSLog(@"Failed");
            
        });
    });
}


@end

File: MDB_Swift/MyDailyBeat/MyDailyBeat/src/EVCGroupSettingsViewController.swift

//
//  EVCGroupSettingsViewController.swift
//  MyDailyBeat
//
//  Created by Virinchi Balabhadrapatruni on 10/26/14.
//  Copyright (c) 2014 eVerveCorp. All rights reserved.
//

import UIKit
import API
import Toast_Swift
import FXForms
class EVCGroupSettingsViewController: UITableViewController {
    var frm = GroupPrefs()
    var group = Group()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.navigationItem.title = "Edit Group"
        self.tableView.register(ToggleTableViewCell.self, forCellReuseIdentifier: "HobbiesCell")
        self.tableView.register(UITableViewCell.self, forCellReuseIdentifier: "DeleteCell")
        let saveButton = UIBarButtonItem(barButtonSystemItem: .save, target: self, action: #selector(editGroup))
        self.navigationItem.rightBarButtonItem = saveButton
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self.tableView.reloadData()
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        self.loadPrefs()
    }
    
    func loadPrefs() {
        DispatchQueue.global().async(execute: {() -> Void in
            DispatchQueue.main.async(execute: {() -> Void in
                self.view.makeToastActivity(.center)
            })
            let hobbies = RestAPI.getInstance().getHobbiesForGroup(self.group).map({ (json) -> Int in
                return json["hby_ref_id"].intValue
            })
            
            for key in HobbiesRefList.getInstance().list.keys {
                self.frm.hobbies[key] = hobbies.contains(key)
            }
            
            DispatchQueue.main.async(execute: {() -> Void in
                self.view.hideToastActivity()
                self.tableView.reloadData()
            })
        })
    }
    
    
    func deleteGroup() {
        DispatchQueue.global().async(execute: {() -> Void in
            DispatchQueue.main.async(execute: {() -> Void in
                self.view.makeToastActivity(.center)
            })
            let success: Bool = RestAPI.getInstance().delete(group: self.group)
            DispatchQueue.main.async(execute: {() -> Void in
                self.view.hideToastActivity()
                if success {
                    self.view.makeToast("Group delete successful!", duration: 3.5, position: .bottom)
                }
                else {
                    self.view.makeToast("Group delete failed!", duration: 3.5, position: .bottom)
                    return
                }
                self.navigationController?.setNavigationBarHidden(false, animated: true)
                self.navigationController?.popToRootViewController(animated: true)
            })
        })
    }
    
    func editGroup() {
        DispatchQueue.global().async(execute: {() -> Void in
            DispatchQueue.main.async(execute: {() -> Void in
                self.view.makeToastActivity(.center)
            })
            let success: Bool = RestAPI.getInstance().setHobbiesforGroup(ID: self.group.groupID, self.frm)
            DispatchQueue.main.async(execute: {() -> Void in
                self.view.hideToastActivity()
                if success {
                    self.view.makeToast("Group editing successful!", duration: 3.5, position: .bottom)
                }
                else {
                    self.view.makeToast("Group editing failed!", duration: 3.5, position: .bottom)
                    return
                }
                self.navigationController?.setNavigationBarHidden(false, animated: true)
                self.navigationController?.popToRootViewController(animated: true)
            })
        })
    }
    
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 2
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if section == 1 {
            return 1
        } else {
            return HobbiesRefList.getInstance().list.count
        }
    }
    
    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        if section == 1 {
            return nil
        } else {
            return "Group Hobbies"
        }
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        if indexPath.section == 1 {
            let textCell = tableView.dequeueReusableCell(withIdentifier: "DeleteCell", for: indexPath)
            textCell.textLabel?.textAlignment = .center
            textCell.textLabel?.textColor = UIColor.white
            textCell.backgroundColor = UIColor.red
            textCell.textLabel?.text = "Delete Group"
            return textCell
        } else {
            let toggleCell = tableView.dequeueReusableCell(withIdentifier: "HobbiesCell", for: indexPath) as! ToggleTableViewCell
            let list = HobbiesRefList.getInstance()
            toggleCell.textLabel?.text = list.list[indexPath.row + 1]!
            toggleCell.update()
            toggleCell.toggleSwitch.setOn(self.frm.hobbies[indexPath.row + 1] ?? false, animated: true)
            toggleCell.onToggle = {
                self.frm.toggle(key: indexPath.row + 1, value: toggleCell.toggleSwitch.isOn)
                self.tableView.reloadData()
            }
            return toggleCell
        }
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        if indexPath.section == 1 {
            self.deleteGroup()
        }
    }
    
    
}

